client/src/components/Filter.jsx

```diff
@@ -10,8 +24,208 @@ import React from "react";
+    });
+
+    if (hasAtLeastOneChecked) changeButtonNames(1);
+    else changeButtonNames(0);
+    handlePopUp(true, event.currentTarget);
+  };
+
+  const handleClose = () => {
+    handlePopUp(false);
+    setChecked(allFilterData[filterLabel].data);
+  };
+
+  const handleChange = (event) => {
+    setChecked({
+      ...checked,
+      [event.target.name]: {
+        ...checked[event.target.name],
+        checked: event.target.checked,
+      },
+    });
+    if (buttonStatus > 0) changeButtonNames(2);
+  };
+
+  const handleSubmit = () => {
+    setAllFilterData({
+      ...allFilterData,
+      [filterLabel]: {
+        ...allFilterData[filterLabel],
+        data: {
+          ...checked,
+        },
+      },
+    });
+    handlePopUp(false);
+  };
+
+  const handleCancel = () => {
+    if (buttonStatus === 2) {
+      setChecked(allFilterData[filterLabel].data);
+      changeButtonNames(1);
+    } else {
+      setChecked(
+        Object.keys(checked).map(
+          (key) => (checked[key] = { ...checked[key], checked: false })
+        )
+      );
+      setButtonStatus(0);
+    }
+  };
+
+  const changeButtonNames = (status) => {
+    switch (status) {
+      case 0:
+        setCancelButton("Cancel filter");
+        break;
+      case 1:
+        setCancelButton("Reset filter");
+        break;
+      default:
+        setCancelButton("Cancel changes");
+    }
+    setButtonStatus(status);
+  };
+
+  const open = Boolean(anchorEl);
+  const id = open ? allFilterData[filterLabel].filterName : undefined;
+  const popOverProps = {
+    sx: {
+      backgroundColor: "gray.50",
+      boxShadow:
+        "0px 1px 1px rgba(0, 0, 0, 0.14), 0px 1px 2px rgba(0, 0, 0, 0.12), 0px 1px 3px rgba(0, 0, 0, 0.2)",
+      marginTop: "0.5rem",
+      borderRadius: "8px",
+    },
+  };
+  return (
+    <>
+      <Button
+        aria-describedby={id}
+        onClick={handleClick}
+        endIcon={icon}
+        variant="contained"
+        sx={{
+          bgcolor: buttonBgColor,
+          border: "1px solid",
+          borderColor: buttonBorderColor,
+          boxShadow: "none",
+          borderRadius: "8px",
+          textTransform: "none",
+          padding: ".3rem 1rem",
+          color: buttonBorderColor,
+          "&:hover": {
+            bgcolor: buttonBgColor,
+            boxShadow: "none",
+          },
+        }}
+      >
+        <Typography noWrap sx={{ fontSize: "1rem" }}>
+          {allFilterData[filterLabel].filterName}
+        </Typography>
+      </Button>
+      <Popover
+        id={id}
+        open={open}
+        anchorEl={anchorEl}
+        onClose={handleClose}
+        anchorOrigin={{
+          vertical: "bottom",
+          horizontal: "left",
+        }}
+        PaperProps={popOverProps}
+      >
+        <FormControl sx={{ mx: 3, my: 1 }}>
+          <FormGroup>
+            {Object.keys(checked).map((key) => (
+              <Grid container key={key} alignItems="center">
+                <Grid item xs>
```

Comment:
Just out of curiosity, do you have a preference over grid vs flex? Personally i don't find too many instances where grid is helpful for much smaller components like this, and usually more helpful forcing columns on larger layouts.

Not saying this is wrong, just Grid adds so many elements vs a few styling options for flex containers and flex items.

Reply:
ngl I completely forgot about the flex attribute...once I started MUI I kind of replaced flex with Grid...would you like me to use flex here instead to shorten the code?

Reply:
i think up to you, i dont see a problem with flex in this situation, and if grid is working, thats fine too. Grid just creates the need for more elements to organize spaces so small, and i think columns becomes tough to work with at that point. But also if we're not using columns, maybe we just use flex